modules/tinymce/src/core/main/ts/fmt/NormalizeTagOrder.ts

```diff
@@ -0,0 +1,164 @@
+import * as RemoveFormat from './RemoveFormat';
+
+interface WrapperData {
+  readonly container: SugarElement<Element>;
+  readonly innerWrapper: SugarElement<Element>;
+  readonly outerWrappers: SugarElement<Element>[];
+}
+
+const fontSizeAlteringFormats = [ 'fontsize', 'subscript', 'superscript' ] as const;
+const formatsToActOn = [ 'strikethrough', ...fontSizeAlteringFormats ] as const;
+
+const hasFormat = (formatter: Formatter, el: SugarElement<Node>, format: string): el is SugarElement<Element> =>
+  Type.isNonNullable(formatter.matchNode(el.dom, format, {}, format === 'fontsize'));
+
+const isFontSizeAlteringElement = (formatter: Formatter) => (el: SugarElement<Node>): el is SugarElement<Element> =>
+  Arr.exists(fontSizeAlteringFormats, (format) => hasFormat(formatter, el, format));
+
+const isFontSizeAlteringFormat = (format: string) => Arr.contains(fontSizeAlteringFormats, format);
+
+const isNormalizingFormat = (format: string) => Arr.contains(formatsToActOn, format);
+
+const gatherWrapperData = (
+  isRoot: (node: SugarElement<Node>) => boolean,
+  scope: SugarElement<Element>,
+  hasFormat: (el: SugarElement<Element>) => boolean,
+  createFormatElement: (el: SugarElement<Element>) => SugarElement<Element>,
+  removeFormatFromElement: (el: SugarElement<Element>) => Optional<SugarElement<Element>>
+): Optional<WrapperData> => {
+  const parents = Traverse.parents(scope, isRoot).filter(SugarNode.isElement);
+
+  return Arr.findLastIndex(parents, hasFormat).map((index) => {
+    const container = parents[index];
+    const innerWrapper = createFormatElement(container);
+    const outerWrappers = [
+      ...removeFormatFromElement(Replication.shallow(container)).toArray(),
+      ...Arr.bind(parents.slice(0, index), (wrapper) => {
+        if (hasFormat(wrapper)) {
+          return removeFormatFromElement(wrapper).toArray();
+        } else {
+          return [ Replication.shallow(wrapper) ];
+        }
+      })
+    ];
+
+    return { container, innerWrapper, outerWrappers };
+  });
+};
+
+const wrapChildrenInInnerWrapper = (
+  target: SugarElement<Element>,
+  wrapper: SugarElement<Element>,
+  hasFormat: (el: SugarElement<Node>) => boolean,
+  removeFormatFromElement: (el: SugarElement<Element>) => Optional<SugarElement<Element>>
+) => {
+  Arr.each(Traverse.children(target), (child) => {
+    if (SugarNode.isElement(child) && hasFormat(child)) {
+      if (removeFormatFromElement(child).isNone()) {
+        Remove.unwrap(child);
+      }
+    }
+  });
+
+  Arr.each(Traverse.children(target), (child) => Insert.append(wrapper, child));
+
+  Insert.prepend(target, wrapper);
+};
+
+const wrapInOuterWrappers = (target: SugarElement<Element>, wrappers: SugarElement<Element>[]) => {
+  if (wrappers.length > 0) {
+    const outermost = wrappers[wrappers.length - 1];
+    Insert.before(target, outermost);
+
+    const innerMost = Arr.foldl(wrappers.slice(0, wrappers.length - 1), (acc, wrapper) => {
```

Comment:
Why do you need to copy the array here? I don't think you're modifying it 🤔

Reply:
It's a copy of everything but the last element.